### ISSUE TITLE
win_regedit: change detection doesn't work for environment variables in registry keys of type expandstring

### ISSUE THREAD

**Issue by chrwen-omicron** [ansible/ansible-modules-extras#2446]
##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

win_regedit
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### SUMMARY

Setting an environment variable as data of an expandstring data type leads to a change every run.
##### STEPS TO REPRODUCE

```
- name: set an environment variable as data
  win_regedit:
    key: "HKCU:\\Environment"
    value: HOME
    data: "%USERPROFILE%"
    type: expandstring
```

##### EXPECTED RESULTS

No change if the registry contains `%USERPROFILE%.`
##### ACTUAL RESULTS

When win_regedit is reading the registry key, it gets the expanded data instead of the real data.
E.g. `C:\Users\<username>` instead of `%USERPROFILE%`

**Comment by jhawkesworth** [comment]
A change was merged into devel yesterday that affected change detection for win_regedit.
Please could you try latest devel?

**Comment by chrwen-omicron** [comment]
Sorry for not mentioning this, I did try the most recent version of win_regedit in the devel branch.

The problem seems to be `Get-ItemProperty` in

``` powershell
$currentRegistryData =(Get-ItemProperty -Path $registryKey | Select-Object -ExpandProperty $registryValue)
```

This returns the expanded string.

**Comment by jhawkesworth** [comment]
thanks @chrwen-omicron - Thinking about it further, I don't have any occurrences of expand string in my playbooks right now, and I'm pretty certain this isn't covered in the existing integration tests.

I suspect the type of $currentRegistryData is not one that is being handled in Compare-RegistryData

Could you find the type of $currentRegistryData when working with an expandstring?

```
write-host $currentRegistryData.GetType().Name 
```

would get it.  Might be easiest to set ANSIBLE_KEEP_REMOTE_FILES=1 on your controller, run the playbook once then re-run in Powershell ISE on a windows node otherwise write-host likely would get lost.

If it isn't a String, int or Object[] then the compare test won't handle it at the moment.

```
        if ($ReferenceData -is [String] -or $ReferenceData -is [int]) {
            if ($ReferenceData -eq $DifferenceData) {
                return $true
            } else {
                return $false
            }
        } elseif ($ReferenceData -is [Object[]]) {
            if (@(Compare-Object $ReferenceData $DifferenceData -SyncWindow 0).Length -eq 0) {
                return $true
            } else {
                return $false
            }
        }
```

If it is one of the existing types then its possible we are not comparing the right thing - because what we are storing in the ansible configuration isn't the final value, just something that can be expanded out to the right value, we might not be able to make a valid comparison to see whether it has changed.

Also, win_environment exists if you want to manage environment variables.

Hope this helps,

Jon

**Comment by chrwen-omicron** [comment]
Well, the registry types REG_SZ and REG_EXPAND_SZ seem to be basically the same. They're both strings, so the compare test is fine, but REG_EXPAND_SZ may contain unexpanded references to environment variables.
`if ($ReferenceData -is [String] -or $ReferenceData -is [int])` is handling it correctly, but the `$ReferenceData` is expanded and the `$DifferenceData` not, so they differ at `if ($ReferenceData -eq $DifferenceData)`.

I tried to use win_environment in the first place, but ended up using win_regedit because it doesn't support expandstring.

This line seems to do the trick and returns `%USERPROFILE%` instead of `C:\Users\<username>`:

``` powershell
(Get-Item -Path $registryKey).GetValue($registryValue, "", "DoNotExpandEnvironmentNames")
```

But I don't know powershell, so this might have side effects which I'm not aware of.

Regards

**Comment by ansibot** [comment]
@joshludwig, @smadam813, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@joshludwig, @smadam813, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jhawkesworth** [comment]
I think I am against making a change to support this as it would mean that you couldn't look at your ansible config and determine ahead of time for certain what the final registry value would be. I can see how it is desirable from the point of view of being automatable, but it stops ansible configuration being a 'single source of truth', which I value. I get that it can make it harder to accommodate the many and varied needs of your system users.

@chrwen-omicron - 
Perhaps you could give an example use case to make it easier to reason about?

**Comment by chrwen-omicron** [comment]
@jhawkesworth following your argument you're not allowed to use environment variables at all in your ansible config.

The important thing here is that the write operation behaves differently from the read operation. The write can handle unexpanded environment variables for the REG_EXPAND_SZ type whereas the read can't.
At least you should get an option to select the behavior.

This is also a problem if you want to extend your PATH variable for instance, because you're expanding the variables inside the PATH.

**Comment by ansibot** [comment]
@joshludwig, @smadam813, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@joshludwig, @smadam813, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@joshludwig, @smadam813, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@joshludwig, @smadam813, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
This repository has been locked. All new issues and pullrequests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pullrequest to the ansible/ansible repo.

**Comment by chrwen-omicron** [comment]
This issue was moved to ansible/ansible#19213
